lib/internal/modules/esm/default_resolve.js

```diff
@@ -51,17 +50,41 @@ if (experimentalWasmModules)
 if (experimentalJsonModules)
   extensionFormatMap['.json'] = legacyExtensionFormatMap['.json'] = 'json';
 
+function getFormat({ url, isMain }) {
+  if (NativeModule.canBeRequiredByUsers(url)) {
+    return { format: 'builtin' };
+  }
+  const parsed = new URL(url);
+  if (parsed.protocol === 'data:') {
+    const [ , mime ] = /^([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/.exec(parsed.pathname) || [ null, null, null ];
+    const format = ({
+      '__proto__': null,
+      'text/javascript': 'module',
+      'application/json': 'json',
+      'application/wasm': experimentalWasmModules ? 'wasm' : null
+    })[mime] || null;
+    return { format };
+  } else if (parsed.protocol === 'file:') {
+    const ext = extname(parsed.pathname);
+    let format = extensionFormatMap[ext];
+    if (ext === '.js' || (!format && isMain))
+      format = getPackageType(url.href) === TYPE_MODULE ? 'module' : 'commonjs';
+    if (!format && experimentalSpeciferResolution === 'node') {
+      process.emitWarning(
+        'The Node.js specifier resolution in ESM is experimental.',
+        'ExperimentalWarning');
+      format = legacyExtensionFormatMap[ext];
+    }
+    return { format: format || null };
```

Comment:
Do we need to make this an object, or can we just permit a string return?

Reply:
I guess it depends on it we might extend this in the future to return more data? Perhaps data or functions to pass between hooks?

Reply:
For `getFormat` in particular I feel that this should be an object while we sort out how things like bodies can be ferried between to avoid double parsing/fetching.

Reply:
Can we perhaps still support the direct string as a sugar case though?

Reply:
I'm not the biggest fan of polymorphic return types since that means each stage has to do branching marshaling to/from strings. If we keep it the same as the value returned by `getDefaultFormat` that seems better than branching to me. I could be swayed otherwise but would rather start throwing on invalid returns like the documentation example has/had rather than trying to marshal multiple types. I think an alternative of moving `getFormat` to be combined with `getSource` would remove the need for this data ferrying as well and might be more palpable to myself.